worlds/ahit/Regions.py

```diff
@@ -0,0 +1,947 @@
+    i = 1
+    for name in rift_access_regions[region.name]:
+        act_region = world.multiworld.get_region(name, world.player)
+        entrance_name = f"{region.name} Portal - Entrance {i}"
+        connect_regions(act_region, region, entrance_name, world.player)
+        i += 1
+
+    # fix for some weird keyerror from tests
+    if region.name == "Time Rift - Rumbi Factory":
+        for entrance in region.entrances:
+            world.multiworld.get_entrance(entrance.name, world.player)
+
+
+def create_tasksanity_locations(world: "HatInTimeWorld"):
+    ship_shape: Region = world.multiworld.get_region("Ship Shape", world.player)
+    id_start: int = TASKSANITY_START_ID
+    for i in range(world.options.TasksanityCheckCount.value):
+        location = HatInTimeLocation(world.player, f"Tasksanity Check {i+1}", id_start+i, ship_shape)
+        ship_shape.locations.append(location)
+
+
+def is_valid_plando(world: "HatInTimeWorld", region: str, is_candidate: bool = False) -> bool:
+    # Duplicated keys will throw an exception for us, but we still need to check for duplicated values
+    if is_candidate:
+        found_list: List = []
+        old_region = region
+        for name in world.options.ActPlando.keys():
+            act = world.options.ActPlando.get(name)
+            if act == old_region:
+                region = name
+                found_list.append(name)
+
+        if len(found_list) == 0:
+            return False
+
+        if len(found_list) > 1:
+            raise Exception(f"ActPlando ({world.multiworld.get_player_name(world.player)}) - "
+                  f"Duplicated act plando mapping found for act: \"{old_region}\"")
+    elif region not in world.options.ActPlando.keys():
+        return False
+
+    if region in blacklisted_acts.values() or (region not in act_entrances.keys() and "Time Rift" not in region):
+        return False
+
+    act = world.options.ActPlando.get(region)
+    try:
+        world.multiworld.get_region(region, world.player)
+        world.multiworld.get_region(act, world.player)
+    except KeyError:
+        return False
+
+    if act in blacklisted_acts.values() or (act not in act_entrances.keys() and "Time Rift" not in act):
+        return False
+
+    # Don't allow plando-ing things onto the first act that aren't completable with nothing
+    is_first_act: bool = act_chapters[region] == get_first_chapter_region(world).name \
+        and region in act_entrances.keys() and ("Act 1" in act_entrances[region] or "Free Roam" in act_entrances[region])
+
+    if is_first_act:
+        if act_chapters[act] == "Subcon Forest" and world.options.ShuffleSubconPaintings.value > 0:
+            return False
+
+        if world.options.UmbrellaLogic.value > 0 \
+           and (act == "Heating Up Mafia Town" or act == "Queen Vanessa's Manor"):
+            return False
+
+        if act not in guaranteed_first_acts:
+            return False
+
+    # Don't allow straight up impossible mappings
```

Comment:
The thing is, some of the entries in `blacklisted_combos` aren't actually impossible, but are just there to prevent the possibility of ActRando generating an impossible layout (such as Bon Voyage being blacklisted on The Owl Express). I figured that I would only include the *actually* impossible mappings here to allow a bit more freedom with ActPlando combinations.